netconf.c

```diff
@@ -853,8 +864,413 @@ get_full_tree ()
+    {
+        next_node = cur_node->next;
+        if (cur_node->type == XML_ELEMENT_NODE) {
+            if (!g_hash_table_lookup (node_table, cur_node))
+            {
+                xmlUnlinkNode (cur_node);
+                xmlFreeNode (cur_node);
+                if (depth == 0 && cur_node == node)
+                    *root_deleted = true;
+
+                continue;
+            }
+        }
+
+        cleanup_xpath_tree (node_table, cur_node->children, depth + 1, max_depth, root_deleted);
+    }
+    if (depth > *max_depth)
+        *max_depth = depth;
+
+    if (depth == 0)
+        cleanup_empty_branches (node, depth, *max_depth, root_deleted);
+}
+
+void
+xpath_tree_add (GHashTable *node_table, xmlNode *node)
+{
+    xmlNode *cur_node = NULL;
+    xmlNode *next_node = NULL;
+
+    for (cur_node = node; cur_node; cur_node = next_node) {
+        next_node = cur_node->next;
+        if (cur_node->type == XML_ELEMENT_NODE)
+        {
+            if (!g_hash_table_lookup (node_table, cur_node))
+                g_hash_table_insert (node_table, cur_node, cur_node);
+        }
+
+        xpath_tree_add (node_table, cur_node->children);
+    }
+}
+
+static char *
+prepare_xpath_eval_path (char *path, char *ns_prefix)
+{
+    char *xpath;
+    char *new_path = NULL;
+
+    if (path[0] == '/')
+    {
+        if (strlen (path) > 1 && path[1] != '/')
+        {
+            char *colon;
+            colon = strchr (path + 1, ':');
+            if (!colon && ns_prefix)
+            {
+                new_path = g_strdup_printf ("/%s:%s", ns_prefix, path + 1);
+                path = new_path;
+            }
+        }
+    }
+
+    xpath = g_strdup (path);
+    g_free (new_path);
+    return xpath;
+}
+
+void
+xpath_set_namespace (char *path, char **ns_href, char **ns_prefix, xmlDoc *doc,
+                     xmlNode *xml, xmlXPathContext *xpath_ctx)
+{
+    char *href = NULL;
+    char *prefix = NULL;
+    char *next;
+    char *top_node;
+
+    if (!*ns_href || !*ns_prefix)
+    {
+        if (xml->ns)
+        {
+            if (xml->ns->href)
+            {
+                href = g_strdup ((char *) xml->ns->href);
+                if (*ns_href)
+                    g_free (*ns_href);
+                *ns_href = href;
+            }
+            if (xml->ns->prefix)
+            {
+                prefix = g_strdup ((char *) xml->ns->prefix);
+                if (*ns_prefix)
+                    g_free (*ns_prefix);
+                *ns_prefix = prefix;
+            }
+        }
+        if (!*ns_href || !*ns_prefix)
+        {
+            if (path[0] == '/')
+            {
+                next = strchr (path + 1, '/');
+                top_node = g_strndup (path + 1, next - path - 1);
+                xmlNode *lnode = sch_lookup (g_schema, top_node);
+                href = NULL;
+                prefix = NULL;
+                if (lnode->ns)
+                {
+                    if (lnode->ns->href)
```

Comment:
How are we able to use href and prefix as if they are allocated bits of memory?

Reply:
If the schema node has a namespace set then set href and prefix from that information. Sometimes the schema node only has one of the required fields, hence if both are not present, then try getting the namespace information from another source

Reply:
Sorry, I was querying whether getting href and prefix in the way you do (seems to be pointers to existing memory) was OK given that there's a possibility that they are freed later on.

Reply:
OK, I see you did fix it.

Reply:
Or did you - href and prefix above this line are strduped. but immediately after this line, they are assigned and not duplicated.